do/csi/driver/controller.go

```diff
@@ -3,48 +3,347 @@ package driver
+// idempotent.
+func (d *Driver) CreateVolume(ctx context.Context, req *csi.CreateVolumeRequest) (*csi.CreateVolumeResponse, error) {
+	if req.Name == "" {
+		return nil, status.Error(codes.InvalidArgument, "CreateVolume Name must be provided")
+	}
+
+	if req.VolumeCapabilities == nil || len(req.VolumeCapabilities) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "CreateVolume Volume capabilities must be provided")
+	}
+
+	size, err := extractStorage(req.CapacityRange)
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	volumeName := req.Name
+
+	// get volume first, if it's created do no thing
+	volumes, _, err := d.doClient.Storage.ListVolumes(ctx, &godo.ListVolumeParams{
+		Region: d.region,
+		Name:   volumeName,
+	})
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	// volume already exist, do not thing
+	if len(volumes) != 0 {
+		if len(volumes) > 1 {
+			return nil, fmt.Errorf("fatal issue: duplicate volume %q exists", volumeName)
+		}
+		vol := volumes[0]
+
+		// check if it was created by the CSI driver
+		if vol.Description != createdByDO {
+			return nil, fmt.Errorf("fatal issue: volume %q (%s) was not created by CSI",
+				vol.Name, vol.Description)
+		}
+
+		if vol.SizeGigaBytes*GB != size {
+			return nil, status.Error(codes.AlreadyExists, fmt.Sprintf("invalid option requested size: %d", size))
+		}
+
+		return &csi.CreateVolumeResponse{
+			Volume: &csi.Volume{
+				Id:            vol.ID,
+				CapacityBytes: vol.SizeGigaBytes * GB,
+			},
+		}, nil
+	}
+
+	volumeReq := &godo.VolumeCreateRequest{
+		Region:        d.nodeId,
+		Name:          volumeName,
+		Description:   createdByDO,
+		SizeGigaBytes: size / GB,
+	}
+
+	// TODO(arslan): Currently DO only supports SINGLE_NODE_WRITER mode. In the
+	// future, if we support more modes, we need to make sure to create a
+	// volume that aligns with the incoming capability. We need to make sure to
+	// test req.VolumeCapabilities
+	vol, _, err := d.doClient.Storage.CreateVolume(ctx, volumeReq)
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	return &csi.CreateVolumeResponse{
+		Volume: &csi.Volume{
+			Id:            vol.ID,
+			CapacityBytes: size,
+		},
+	}, nil
 }
 
-// DeleteVolume deletes the given volume. This doesn't mean it needs to be
-// purged away. It needs to be deprovisioned, which means it should not be
-// accesible by anymore.
-func (d *Driver) DeleteVolume(context.Context, *csi.DeleteVolumeRequest) (*csi.DeleteVolumeResponse, error) {
-	return nil, errors.New("not implemented")
+// DeleteVolume deletes the given volume. The function is idempotent.
+func (d *Driver) DeleteVolume(ctx context.Context, req *csi.DeleteVolumeRequest) (*csi.DeleteVolumeResponse, error) {
+	if req.VolumeId == "" {
+		return nil, status.Error(codes.InvalidArgument, "DeleteVolume Volume ID must be provided")
+	}
+
+	_, err := d.doClient.Storage.DeleteVolume(ctx, req.VolumeId)
```

Comment:
Not familiar enough with CSI yet to know what is calling this method but do we need to check volume description here?

Reply:
Here is the full logic step by step (From https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/container-storage-interface.md):

<img width="900" alt="screen shot 2018-04-08 at 12 15 40 am" src="https://user-images.githubusercontent.com/438920/38460420-01db8c3c-3ac2-11e8-94b9-7fbcc57776db.png">

In this case, because the Volume is bound to the PVC, I'm thinking how we can accidentally deleted it? Because we use the Volume ID that is bound to the PVC.